components/places/src/storage.rs

```diff
@@ -263,21 +270,314 @@ fn add_visit(db: &impl ConnExt,
+            Some(info) => {
+                assert_eq!(info.page.guid, *new_guid);
+                info.page
+            },
+            None => {
+                assert!(existing_guid.is_none());
+                new_page_info(db, &url, Some(new_guid.clone()))?
+            },
+        };
+        for visit in visits {
+            let transition = VisitTransition::from_primitive(visit.transition).expect("these should already be validated");
+            add_visit(db, &page_info.row_id, &None, &visit.date, &transition, &false)?;
+        }
+        // XXX - we really need a better story for frecency-boost than
+        // Option<bool> - None vs Some(false) is confusing. We should use an enum.
+        update_frecency(&db, page_info.row_id, None)?;
+
+        // and the place itself if necessary.
+        let new_title: &String = match title {
+            Some(title) => title,
+            None => &page_info.title,
+        };
+        // We set the Status to Normal, otherwise we will re-upload it as
+        // outgoing even if nothing has changed.
+        db.execute_named_cached(
+            "UPDATE moz_places
+             SET title = :title,
+                 sync_status = :status
+             WHERE id == :row_id",
+            &[(":title", new_title),
+              (":row_id", &page_info.row_id),
+              (":status", &SyncStatus::Normal)]
+        )?;
+
+        // and finally update the guid if necessary.
+        // XXX - we should consider doing this as part of the update above, but
+        // it complicates the code and should be rare.
+        if let Some(ref existing_guid) = existing_guid {
+            if existing_guid != new_guid {
+                db.execute_named_cached(
+                    "UPDATE moz_places SET guid = :new_guid WHERE guid == :old_guid",
+                    &[(":new_guid", new_guid), (":old_guid", existing_guid)]
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn apply_synced_reconcilliation(db: &Connection, guid: &SyncGuid) -> Result<()> {
+        db.execute_named_cached(
+            "UPDATE moz_places
+             SET sync_status = :status,
+                 sync_change_counter = 0
+             WHERE guid == :guid",
+            &[(":guid", guid),
+              (":status", &SyncStatus::Normal)]
+        )?;
+        Ok(())
+    }
+
+    pub fn apply_synced_deletion(db: &Connection, guid: &SyncGuid) -> Result<()> {
+        db.execute_named_cached(
+            "DELETE FROM moz_places WHERE guid = :guid",
+            &[(":guid", guid)])?;
+        // This might have created a local tombstone for the item, which we
+        // don't want.
+        db.execute_named_cached(
+            "DELETE FROM moz_places_tombstones WHERE guid = :guid",
+            &[(":guid", guid)])?;
+        Ok(())
+    }
+
+    #[derive(Debug)]
+    pub enum OutgoingInfo {
+        Record(HistoryRecord),
+        Tombstone,
+    }
+
+    pub fn fetch_outgoing(db: &impl ConnExt,
```

Comment:
Does &Connection work here (and below)? I don't remember why I did the first &impl ConnExt, but I think it's probnably always a mistake, and we should just use &Connection.